framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -350,33 +483,86 @@ private void refineArrayAtLengthAccess(FieldAccessNode arrayLengthNode, CFStore
         return stringConcatenation(n.getLeftOperand(), n.getRightOperand(), p, result);
     }
 
-    public TransferResult<CFValue, CFStore> stringConcatenation(
-            Node leftOperand,
-            Node rightOperand,
-            TransferInput<CFValue, CFStore> p,
-            TransferResult<CFValue, CFStore> result) {
-        List<String> lefts = getStringValues(leftOperand, p);
-        List<String> rights = getStringValues(rightOperand, p);
-        List<String> concat;
-        if (lefts == null || rights == null) {
-            concat = null;
-        } else {
-            concat = new ArrayList<>();
-            if (lefts.isEmpty()) {
-                lefts = Collections.singletonList("null");
+    /**
+     * Calculates possible lengths of a result of string concatenation of strings with known
+     * lengths.
+     */
+    private List<Integer> calculateLengthAddition(
+            List<Integer> leftLengths, List<Integer> rightLengths) {
+        ArrayList<Integer> result = new ArrayList<Integer>();
+
+        for (int left : leftLengths) {
+            for (int right : rightLengths) {
+                long resultLength = left + right;
+                // Lengths not fitting into int are not allowed
+                if (resultLength <= Integer.MAX_VALUE) result.add((int) resultLength);
             }
-            if (rights.isEmpty()) {
-                rights = Collections.singletonList("null");
+        }
+
+        return result;
+    }
+
+    /**
+     * Calculates a range of possible lengths of a result of string concatenation of strings with
+     * known ranges of lengths.
+     */
+    private Range calculateLengthRangeAddition(Range leftLengths, Range rightLengths) {
+        return leftLengths.plus(rightLengths).intersect(Range.INT_EVERYTHING);
+    }
+
+    private AnnotationMirror createAnnotationForStringConcatenation(
```

Comment:
Comments describing the different sections of this method (i.e. first check if both are StringVals, then try lengths, etc.) would make it easier to read.